### PR TITLE
HW-23 DAO-8

### PR DESCRIPTION
**Pull request by tbondare**

### CODE REVIEW

File: src/main/webapp/web.xml [hunk 1]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>DisplayDrivers</servlet-name>
+        <servlet-class>taxiapp.controller.GetAllDriversController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>DisplayDrivers</servlet-name>
+        <url-pattern>/drivers/display</url-pattern>
```
Comment (oroboro-cpu) [inline src/main/webapp/web.xml:24]: Don’t use plural for package names.
```suggestion
        <url-pattern>/driver/display</url-pattern>
```

File: src/main/webapp/web.xml [hunk 2]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>createManufacturer</servlet-name>
+        <servlet-class>taxiapp.controller.CreateManufacturerController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>createManufacturer</servlet-name>
+        <url-pattern>/manufacturers/create</url-pattern>
```
Comment (oroboro-cpu) [inline src/main/webapp/web.xml:44]: Same here
```suggestion
        <url-pattern>/manufacturer/create</url-pattern>
```

File: src/main/webapp/web.xml [hunk 3]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>taxiapp.controller.AddDriverToCarController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/cars/addDriver</url-pattern>
```
Comment (oroboro-cpu) [inline src/main/webapp/web.xml:64]:
```suggestion
        <url-pattern>/cars/drivers/add</url-pattern>
```

File: src/main/java/taxiapp/controller/CreateDriverController.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        req.getRequestDispatcher("/WEB-INF/views/drivers/create.jsp").forward(req, resp);
+
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String licenseNumber = req.getParameter("number");
+        String repeatLicenseNumber = req.getParameter("number-repeat");
```
Comment (oroboro-cpu) [inline src/main/java/taxiapp/controller/CreateDriverController.java:29]: I don't know if it is necessary to make it in this task. We'll implement this in next homework

**Review by oroboro-cpu — CHANGES_REQUESTED** [review]

File: src/main/webapp/WEB-INF/views/cars/create.jsp [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Title</title>
```
Comment (oroboro-cpu) [inline src/main/webapp/WEB-INF/views/cars/create.jsp:4]: Make Title more informative for all jsp files

**Review by oroboro-cpu — CHANGES_REQUESTED** [review]

File: src/main/webapp/web.xml [hunk 6]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>taxiapp.controller.AddDriverToCarController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/addDriver</url-pattern>
```
Comment (oroboro-cpu) [inline src/main/webapp/web.xml:64]:
```suggestion
        <url-pattern>/car/drivers/add</url-pattern>
```

**Review by oroboro-cpu — CHANGES_REQUESTED** [review]

**Review by oroboro-cpu — APPROVED** [review]

File: src/main/java/taxiapp/controller/AddDriverToCarController.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        Long driverId = Long.valueOf(req.getParameter("driverId"));
+        Long carId = Long.valueOf(req.getParameter("carId"));
+        Car car = carService.get(carId);
+        Driver driver = driverService.get(driverId);
+        carService.addDriverToCar(driver, car);
+        req.getRequestDispatcher("/WEB-INF/views/car/driver/add.jsp").forward(req, resp);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/AddDriverToCarController.java:35]:
```suggestion
        resp.sendRedirect(req.getContextPath() + "/cars/drivers/add");
```
let's use sendRedirect here instead

File: src/main/java/taxiapp/controller/CreateCarsController.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import taxiapp.lib.Injector;
+import taxiapp.model.Car;
+import taxiapp.model.Manufacturer;
+import taxiapp.service.CarService;
+import taxiapp.service.ManufacturerService;
+
+public class CreateCarsController extends HttpServlet {
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/CreateCarsController.java:14]:
```suggestion
public class CreateCarController extends HttpServlet {
```

File: src/main/java/taxiapp/controller/CreateDriverController.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        req.getRequestDispatcher("/WEB-INF/views/driver/create.jsp").forward(req, resp);
+
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String licenseNumber = req.getParameter("number");
+        Driver driver1 = new Driver(name, licenseNumber);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/CreateDriverController.java:29]:
```suggestion
        Driver driver = new Driver(name, licenseNumber);
```

File: src/main/java/taxiapp/controller/CreateManufacturerController.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        req.getRequestDispatcher("/WEB-INF/views/manufacturer/create.jsp").forward(req, resp);
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String country = req.getParameter("country");
+
+        Manufacturer manufacturer1 = new Manufacturer(name, country);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/CreateManufacturerController.java:29]:
```suggestion
        Manufacturer manufacturer = new Manufacturer(name, country);
```

File: src/main/java/taxiapp/controller/CreateManufacturerController.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        manufacturerService.create(manufacturer1);
+        System.out.println(name + " " + country);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/CreateManufacturerController.java:31]: let's remove this line

File: src/main/webapp/WEB-INF/views/car/driver/add.jsp [hunk 12]
```diff
@@ -0,0 +1,14 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Add driver to car</title>
+</head>
+<body>
+<form method="post" action="${pageContext.request.contextPath}/car/driver/add">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/driver/add.jsp:7]:
```suggestion
<form method="post" action="${pageContext.request.contextPath}/cars/drivers/add">
```

File: src/main/webapp/WEB-INF/views/driver/create.jsp [hunk 13]
```diff
@@ -0,0 +1,16 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Create driver</title>
+</head>
+<body>
+<h1>Create driver</h1>
+
+<form method="post" action="${pageContext.request.contextPath}/driver/create">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/driver/create.jsp:9]:
```suggestion
<form method="post" action="${pageContext.request.contextPath}/drivers/create">
```

File: src/main/webapp/WEB-INF/views/index.jsp [hunk 14]
```diff
@@ -1,9 +1,15 @@
...
 </head>
 <body>
-Hello world!
+    <h1>Taxi Service</h1>
+
+    <a href="${pageContext.request.contextPath}/driver/create">Create driver </a><br>
+    <a href="${pageContext.request.contextPath}/driver/display">Display drivers </a><br>
+    <a href="${pageContext.request.contextPath}/manufacturer/create">Create manufacturer </a><br>
+    <a href="${pageContext.request.contextPath}/car/create">Create car </a><br>
+    <a href="${pageContext.request.contextPath}/car/driver/add">Add driver to car </a><br>
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/index.jsp:13]:
```suggestion
    <a href="${pageContext.request.contextPath}/drivers/create">Create driver </a><br>
    <a href="${pageContext.request.contextPath}/drivers">Display drivers </a><br>
    <a href="${pageContext.request.contextPath}/manufacturers/create">Create manufacturer </a><br>
    <a href="${pageContext.request.contextPath}/cars/create">Create car </a><br>
    <a href="${pageContext.request.contextPath}/cars/drivers/add">Add driver to car </a><br>
```

File: src/main/webapp/WEB-INF/views/manufacturer/create.jsp [hunk 15]
```diff
@@ -0,0 +1,16 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Create manufacturer</title>
+</head>
+<body>
+<h1>Create manufacturer</h1>
+
+<form method="post" action="${pageContext.request.contextPath}/manufacturer/create">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/manufacturer/create.jsp:9]:
```suggestion
<form method="post" action="${pageContext.request.contextPath}/manufacturers/create">
```

File: src/main/webapp/web.xml [hunk 16]
```diff
@@ -3,15 +3,65 @@
...
     <servlet-mapping>
         <servlet-name>index</servlet-name>
-        <url-pattern>/index</url-pattern>
+        <url-pattern>/</url-pattern>
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>DisplayDrivers</servlet-name>
+        <servlet-class>taxiapp.controller.GetAllDriversController</servlet-class>
+        <load-on-startup>1</load-on-startup>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:20]: let's remove this from all the servlets

File: src/main/webapp/web.xml [hunk 17]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>DisplayDrivers</servlet-name>
+        <url-pattern>/driver/display</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:24]:
```suggestion
        <url-pattern>/drivers</url-pattern>
```

File: src/main/webapp/web.xml [hunk 18]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>createDriver</servlet-name>
+        <servlet-class>taxiapp.controller.CreateDriverController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>createDriver</servlet-name>
+        <url-pattern>/driver/create</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:34]:
```suggestion
        <url-pattern>/drivers/create</url-pattern>
```

File: src/main/webapp/web.xml [hunk 19]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>createManufacturer</servlet-name>
+        <servlet-class>taxiapp.controller.CreateManufacturerController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>createManufacturer</servlet-name>
+        <url-pattern>/manufacturer/create</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:44]:
```suggestion
        <url-pattern>/manufacturers/create</url-pattern>
```

File: src/main/webapp/web.xml [hunk 20]
```diff
@@ -3,15 +3,65 @@
...
-</web-app>
 
+    <servlet>
+        <servlet-name>createCar</servlet-name>
+        <servlet-class>taxiapp.controller.CreateCarsController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>createCar</servlet-name>
+        <url-pattern>/car/create</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:54]:
```suggestion
        <url-pattern>/cars/create</url-pattern>
```

File: src/main/webapp/web.xml [hunk 21]
```diff
@@ -3,15 +3,65 @@
...
+    </servlet-mapping>
+
+    <servlet>
+        <servlet-name>addDriverToCar</servlet-name>
+        <servlet-class>taxiapp.controller.AddDriverToCarController</servlet-class>
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+    <servlet-mapping>
+        <servlet-name>addDriverToCar</servlet-name>
+        <url-pattern>/car/driver/add</url-pattern>
```
Comment (KseniiaMakarova) [inline src/main/webapp/web.xml:64]:
```suggestion
        <url-pattern>/cars/drivers/add</url-pattern>
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/taxiapp/controller/CreateManufacturerController.java [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws ServletException, IOException {
+        String name = req.getParameter("name");
+        String country = req.getParameter("country");
+        Manufacturer manufacturer = new Manufacturer(name, country);
+        manufacturerService.create(manufacturer);
+        System.out.println(name + " " + country);
```
Comment (KseniiaMakarova) [inline src/main/java/taxiapp/controller/CreateManufacturerController.java:30]: remove this line

File: src/main/webapp/WEB-INF/views/manufacturer/create.jsp (hunk 15 above)
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/manufacturer/create.jsp:9]: do you have the controller in web.xml that is mapped to the endpoint `/manufactureR/create`?

File: src/main/webapp/WEB-INF/views/driver/create.jsp (hunk 13 above)
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/driver/create.jsp:9]: same here, will this work?

File: src/main/webapp/WEB-INF/views/car/driver/add.jsp (hunk 12 above)
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/driver/add.jsp:7]: same

File: src/main/webapp/WEB-INF/views/car/create.jsp [hunk 23]
```diff
@@ -0,0 +1,16 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Create car</title>
+</head>
+<body>
+<h1>Create car</h1>
+
+<form method="post" action="${pageContext.request.contextPath}/car/create">
```
Comment (KseniiaMakarova) [inline src/main/webapp/WEB-INF/views/car/create.jsp:9]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/webapp/WEB-INF/views/driver/create.jsp (hunk 13 above)
Comment (tbondare) [reply src/main/webapp/WEB-INF/views/driver/create.jsp:9]: Yes, it is work

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**